File: C_C++_PROJECTS/P_L/hw02/hw02.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstdlib>
using namespace std;

/* Character classes */
#define LETTER 0
#define DIGIT 1
#define UNKNOWN 99

/* Token codes */
#define INT_LIT 10
#define IDENT 11
#define ASSIGN_OP 20
#define ADD_OP 21
#define SUB_OP 22
#define MULT_OP 23
#define DIV_OP 24
#define LEFT_PAREN 25
#define RIGHT_PAREN 26


/* Global declarations */
/* Variables */
int charClass;
char lexeme[100];
char nextChar;
int lexLen;
int token;
int nextToken;
int curr_index;
int cnt_R_paren=0;
int cnt_L_paren=0;

/* Function declarations */
int lookup(char);
void addChar();
char getc();
void getChar();
void getNonBlank();
int lex();
double expr();
double term();
double factor();
double number();

void print_error(){
    cout << "Syntax error!!" << endl;
}

int lookup(char ch){
    switch(ch){
        case '(': addChar(); nextToken = LEFT_PAREN;    cnt_L_paren++;  break; 
        case ')': addChar(); nextToken = RIGHT_PAREN;   cnt_R_paren++;  break;
        case '+': addChar(); nextToken = ADD_OP;        break;
        case '-': addChar(); nextToken = SUB_OP;        break;
        case '*': addChar(); nextToken = MULT_OP;       break;
        case '/': addChar(); nextToken = DIV_OP;        break;
        case '\n': addChar(); nextToken = EOF;           break;
        default : print_error(); exit(0);
    }
    return nextToken;
}

void addChar(){
    if(lexLen <= 98){
        lexeme[lexLen++] = nextChar;
        lexeme[lexLen] = 0;
    }
    else{
        cout << "ERROR - lexeme is too long" << endl;
        exit(0);
    }
}

void getChar(){
    if((nextChar = getc(stdin)) != '\n'){
        if(isalpha(nextChar)){
            print_error();
            exit(0);
        }
        else if(isdigit(nextChar))
            charClass = DIGIT;
            else charClass = UNKNOWN;
    }
    else{
        charClass = EOF;
    }
}

void getNonBlank(){
    while(isspace(nextChar) && nextChar != '\n')
        getChar();
}

int lex(){
    lexLen = 0;
    getNonBlank();
    switch (charClass) {
        /* Parse identifiers */
        case LETTER:
            addChar();
            getChar();
            while (charClass == LETTER || charClass == DIGIT) { addChar(); getChar(); }
            nextToken = IDENT;
            break;
        /* Parse integer literals */
        case DIGIT:
            addChar();
            getChar();
            while (charClass == DIGIT) { addChar(); getChar(); }
            nextToken = INT_LIT;
            break;
        /* Parentheses and operators */
        case UNKNOWN:
            lookup(nextChar);
            getChar();
            break;
        /* EOF */
        case EOF:
            nextToken = EOF;
            lexeme[0] = 'E';
            lexeme[1] = 'O';
            lexeme[2] = 'F';
            lexeme[3] = 0;
            break;
    } /* End of switch */
    #ifdef DEBUG
    printf("Next token is: %d, Next lexeme is %s\n", nextToken, lexeme);
    #endif
    return nextToken;
} /* End of function lex */

double number(){
    int curr_num = stoi(lexeme);
    lex();
    return curr_num;
}

double factor(){
    bool isNegative = false;
    double curr_num;

    if(nextToken == SUB_OP){
        lex();
        isNegative = true;
    }
    if(nextToken == INT_LIT){
        curr_num = number();
    }
    else if(nextToken == LEFT_PAREN){
        lex();
        curr_num = expr();
        if(nextToken == RIGHT_PAREN)
            lex();
        else{
            print_error();
            exit(0);
        }
    }
    else{
        print_error();
        exit(0);
    }
    if(isNegative){
        return curr_num * -1;
    }
    else return curr_num;
}

double term(){
    double curr_num = factor();

    while(nextToken == MULT_OP || nextToken == DIV_OP){
        int curr_Token = nextToken;
        lex();
        if(curr_Token == MULT_OP) curr_num *= factor();
        else curr_num /= factor();

    }
    return curr_num;
}

double expr(){
    double
    curr_num = term();
    
    // if(nextToken == EOF) return curr_num;
    // if(nextToken != ADD_OP && nextToken != SUB_OP){
    //     print_error();
    //     exit(0);
    // }


    while(nextToken == ADD_OP || nextToken == SUB_OP){
        int curr_Token = nextToken;
        lex();
        if(curr_Token == ADD_OP) curr_num += term();
        else curr_num -= term();
    }
    return curr_num;
}

int main(void){
    while(1){
        cout << ">> ";
        getChar();
        lex();

        double ans = expr();

        if(cnt_L_paren != cnt_R_paren){
            print_error();
            exit(0);
        }

        if(nextToken != -1){
            print_error();
            exit(0);
        }

        if((ans - (int)ans) == 0){      // ans = 정수
            cout << ans << endl;
        }
        else{
            cout << fixed;
            cout.precision(6);
            cout << ans <<endl;
        }
    }
    return 0;
}

File: Python_PROJECTS/PL/hw02/hw02.py

# Character classes #
from re import L
import sys
from sqlalchemy import false, true


LETTER = 0
DIGIT = 1
UNKNOWN = 99

# Token codes #
INT_LIT = 10
IDENT = 11
ASSIGN_OP = 20
ADD_OP = 21
SUB_OP = 22
MULT_OP = 23
DIV_OP = 24
LEFT_PAREN = 25
RIGHT_PAREN = 26
EOF = -1

# Global declarations #
# Variables #
charClass = 1
lexeme = [0 for i in range(100)]
nextChar = 'c'
lexLen = 1
token = 1
nextToken = 1
curr_index = 1
cnt_R_paren = 0
cnt_L_paren = 0

def stoi(ch) :
    num=0
    digit=0
    while ch[digit] != 0 :
        digit += 1
    for i in range (1,digit+1) :
        num += (int(ch[digit-i]) * pow(10,i-1))

    return num

def print_error() :
    print("Syntax error!!")

def lookup(ch) :
    global nextToken
    global cnt_L_paren
    global cnt_R_paren
    if ch == '(' :
        addChar()
        nextToken = LEFT_PAREN
        cnt_L_paren += 1
    elif ch == ')':
        addChar()
        nextToken = RIGHT_PAREN
        cnt_R_paren += 1
    elif ch == '+':
        addChar()
        nextToken = ADD_OP
    elif ch == '-':
        addChar()
        nextToken = SUB_OP
    elif ch == '*':
        addChar()
        nextToken = MULT_OP
    elif ch == '/':
        addChar()
        nextToken = DIV_OP
    elif ch == '\n' :
        addChar()
        nextToken = EOF
    else :
        print_error()
        exit()
    return nextToken



def addChar() :
    global lexLen
    if lexLen <= 98 :
        lexeme[lexLen] = nextChar
        lexLen += 1
        lexeme[lexLen] = 0
    else :
        print("ERROR - lexeme is too long")
        exit()

def getChar() :
    global charClass
    global nextChar
    nextChar = sys.stdin.read(1)
    if nextChar != '\n' :
        if nextChar.isalpha() :
            print_error()
            exit()
        elif nextChar.isdigit() :
            charClass = DIGIT
        else : 
            charClass = UNKNOWN
    else :
        charClass = EOF

def getNonBlank() :
    while nextChar.isspace() and nextChar != '\n':
        getChar()

def lex() :
    global nextToken
    global lexLen
    lexLen = 0
    getNonBlank()
    # Parse identifiers #
    if charClass == LETTER :
        addChar()
        getChar()
        while charClass == LETTER or charClass == DIGIT :
            addChar()
            getChar()
    # Parse integer literals #
    elif charClass == DIGIT:
        addChar()
        getChar()
        while charClass == DIGIT :
            addChar()
            getChar()
        nextToken = INT_LIT
    # Parentheses and operators #
    elif charClass == UNKNOWN:
        lookup(nextChar)
        getChar()
    # EOF #
    elif charClass == EOF:
        nextToken = EOF
        lexeme[0] = 'E'
        lexeme[1] = 'O'
        lexeme[2] = 'F'
        lexeme[3] = 0

    # print("Next token is: ",nextToken,", Next lexeme is ",stoi(lexeme))
    return nextToken

def number() :
    curr_num = stoi(lexeme)
    lex()
    return curr_num

def factor() :
    isNegative = False
    curr_num = -1

    if nextToken == SUB_OP :
        lex()
        isNegative = true
    
    if nextToken == INT_LIT:
        curr_num = number()
    elif nextToken == LEFT_PAREN :
        lex()
        curr_num = expr()
        if nextToken == RIGHT_PAREN :
            lex()
        else :
            print_error()
            exit(0)
    else:
            print_error()
            exit(0)
    
    if isNegative :
        return curr_num * -1
    else :
        return curr_num

def term() :
    curr_num = factor()

    while nextToken == MULT_OP or nextToken == DIV_OP :
        curr_Token = nextToken
        lex()
        if curr_Token == MULT_OP :
            curr_num *= factor()
        else :
            curr_num /= factor()
    
    return curr_num

def expr() :
    curr_num = term()

    while nextToken == ADD_OP or nextToken == SUB_OP :
        curr_Token = nextToken
        lex()
        if curr_Token == ADD_OP :
            curr_num += term()
        else :
            curr_num -= term()

    return curr_num

def main() :
    while True :
        print(">> ", end='', flush=True)
        getChar()
        lex()

        ans = expr()

        if cnt_L_paren != cnt_R_paren:
            print_error()
            exit(0)
        else :
            print(ans)
        

# main()
while True :
    print(">> ", end='', flush=True)
    getChar()
    lex()

    ans = expr()

    if cnt_L_paren != cnt_R_paren:
        print_error()
        exit(0)
    else :
        print(ans)




















